caffe2/perfkernels/CMakeLists.txt

```diff
@@ -13,37 +13,63 @@ set(Caffe2_CPU_SRCS ${Caffe2_CPU_SRCS} ${common_srcs})
 
 # We will only build the perf kernel files if the compiler supports avx2
 # extensions.
-# Currently MSVC seems to have a symbol not found error while linking (related
-# to source file order?). As a result we will currently disable the perfkernel
-# in msvc.
-if (NOT MSVC AND CAFFE2_COMPILER_SUPPORTS_AVX2_EXTENSIONS)
-  add_library(Caffe2_perfkernels_avx OBJECT ${avx_srcs})
-  add_library(Caffe2_perfkernels_avx2 OBJECT ${avx2_srcs})
-  add_dependencies(Caffe2_perfkernels_avx Caffe2_PROTO c10)
-  add_dependencies(Caffe2_perfkernels_avx2 Caffe2_PROTO c10)
+if (CAFFE2_COMPILER_SUPPORTS_AVX2_EXTENSIONS)
+  add_library(Caffe2_perfkernels_avx STATIC ${avx_srcs})
+  add_library(Caffe2_perfkernels_avx2 STATIC ${avx2_srcs})
+  add_dependencies(Caffe2_perfkernels_avx Caffe2_PROTO)
+  add_dependencies(Caffe2_perfkernels_avx2 Caffe2_PROTO)
+  target_link_libraries(Caffe2_perfkernels_avx PRIVATE c10)
+  target_link_libraries(Caffe2_perfkernels_avx2 PRIVATE c10)
   if (MSVC)
-    set_target_properties(
-        Caffe2_perfkernels_avx PROPERTIES COMPILE_FLAGS "/arch:AVX")
-    set_target_properties(
-        Caffe2_perfkernels_avx2 PROPERTIES COMPILE_FLAGS "/arch:AVX2")
-    # Currently MSVC doesn't support AVX512
+    target_compile_options(Caffe2_perfkernels_avx
+        PRIVATE "/arch:AVX")
+    target_compile_options(Caffe2_perfkernels_avx2
+        PRIVATE "/arch:AVX2")
```

Comment:
I am not sure how AVX and AVX2 options are defined for MSVC but do we need to add corresponding options for -mfma, mf16c etc.? Are they subsumed in AVX2 for MSVC?

Reply:
Based on what @xkszltl's explanation, unlike -mavx* options in gcc/clang, arch:AVX* options in MSVC only control compiler's auto generation of avx* instructions and defines a few preprocessor symbols. Intrinsics are available regardless. Not sure if __F16C__ and __FMA__ will be defined by /arch:AVX and /arch:AVX2 , respectively. If not, we may want to define them here.

Reply:
__FMA__ is not defined as a warning from cpuinfo (machine has FMA but not build with it) popped out.

Reply:
Thanks for the information. I'll add D__FMA__ to avx2 and D__F16C__ to avx.